app/models/potential_match_builder.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+# A class for building PotentialMatch objects by trying to add pairs of
+# matching prompts.
+class PotentialMatchBuilder
+  attr_accessor :num_prompts_matched, :max_tags_matched
+
+  ALL = -1
+
+  def initialize(request, offer, settings)
+    @request = request
+    @offer = offer
+    @settings = settings
+
+    @num_prompts_matched = 0
+    @max_tags_matched = 0
+  end
+
+  # Check whether the PotentialMatch we're building is valid -- that is,
+  # whether it has the required number of prompt matches.
+  def valid?
+    desired_matches = @settings.num_required_prompts
+    desired_matches = @request.requests.size if desired_matches == ALL
+    @num_prompts_matched >= desired_matches
+  end
+
+  # Check whether the two prompts match, and if so, record information about
+  # the match.
+  def try_prompt_match(request_prompt, offer_prompt)
+    return unless request_prompt.matches?(offer_prompt, @settings)
+    add_prompt_match(request_prompt, offer_prompt)
+  end
+
+  # Record the fact that the two passed-in prompts match.
+  def add_prompt_match(request_prompt, offer_prompt)
+    @num_prompts_matched += 1
+
+    # If we've already matched ALL, we don't need to check the number of
+    # overlapping tags, since we've already got the maximum possible.
+    return if @max_tags_matched == ALL
+
+    # Compute the number of matching tags, and update max_tags_matched if
+    # necessary.
+    curr_tags_matched = request_prompt.count_tags_matched(offer_prompt)
+    @max_tags_matched = [@max_tags_matched, curr_tags_matched].max
+    @max_tags_matched = ALL if curr_tags_matched == ALL
```

Comment:
Occurs to me that using 'ALL' for the purposes of @max_tags_matched privileges a match between short requests/offers. That is, A has R1 with one tag and R2 with seven tags. B has O1 with one tag matching R1. C has O2 with six tags matching R2. Match 1 between A and B will be considered better than match 2 between A and C, even though that one is actually much more specific. Since we only use this if the two prompts are a valid match by requirements, I would be inclined to try and save the absolute number here instead.